editors/code/src/config.ts

```diff
@@ -44,7 +53,7 @@ export class Config {
     private cfg!: vscode.WorkspaceConfiguration;
 
     constructor(private readonly ctx: vscode.ExtensionContext) {
-        vscode.workspace.onDidChangeConfiguration(this.refreshConfig, this, ctx.subscriptions);
+        vscode.workspace.onDidChangeConfiguration(this.onConfigChange, this, ctx.subscriptions);
         this.refreshConfig();
```

Comment:
here is my concern - refreshConfig is sync, onConfigChange is async, fyi

Reply:
Thank you for rising it.
It looks like we use async event handlers here and in the old buggy highlighting module, so this sounds suspicious indeed.

Yet I agree with [this idea](https://github.com/rust-analyzer/rust-analyzer/pull/3666#issuecomment-603600264) for the particular case and willing to try it first.

As for `async` in such cases, @Veetaha definitely knows more about about this so might be worth to discuss it with him.

Reply:
np, just noticed - too much nodejs in my blood to not :)

Reply:
@eriktrom

> in the constructor. the callback `this.onConfigChange`, registered on line 56 by the `vscode.workspace.onDidChangeConfiguration(this.onConfigChange, this, ctx.subscriptions);` event handler is async. But when called, due to being an async callback, we're already in a different turn of the event loop.

This is not entirely the truth. Promises are executed eagerly, but resolved lazily (i.e. at minimum when the current call stack is emptied). The following code proves it
<details/>
<summary>Code itself</summary>

```ts
async function foo() { ++i; }

let i = 0;

foo().then(() => console.log("resolved: " + i));

console.log("After foo(): " + i);
```
</details>

![image](https://user-images.githubusercontent.com/36276403/77562371-99bc7200-6ec8-11ea-8fb4-1cd3688dfbf3.png)

As you see, the variable gets incremented, right after `foo()` because it is executed eagerly (i.e. synchronously).

> So line 57 is being called immediately, and the method this.onConfigChange is non-deterministically called by the event handler (registered on line 56),  I'd need to debug with a breakpoint but a few things can go wrong, the instance could be destroyed in that time period
we might then re-register that async callback if its a hot path, preventing the normal sync stack from flushing, (b/c promises block while they're queue has any microtasks), leading to a timeout, that then requires an host extension restart.

Right, but "*destroyed*" is not the proper term since the lifetime of objects in JavaScript is indeterminate. Properly speaking, the instance can be logically dropped off, but the event listener is not deregistered.

So, the right way to solve it is to make config `implements Disposable` and deregister the event listener when the config is disposed of. *Right now* this is not a big problem.

It may happen only when the user invokes `Restart server` command. Because when doing it we create a new config, but the previous config's event listener is not deregistered.

Anyway, this is not connected with this particular PR at all, but thanks for spotting this!

Reply:
yeah let's not derail - last note, no response needed, sorry 'bout bringing this up!

```
> async function foo() { i++; return; }
undefined
> foo()
Promise { undefined }
> async function foo() { i++; return Promise.resolve(); }
undefined
> foo()
Promise { <pending> } **<-- WARNING, i can haz bug**
```
the first is immediately resolved, the 2nd waits for an event. the callback in question has an if statement that can return early, returning an unwrapped promise. tl;dr

such fun right? over and out